src/configurePipeline/services/source/sourceRepoService.ts

```diff
@@ -0,0 +1,172 @@
+     */
+    public addYmlFileToRepo(ymlFilePath: string, context: any) {
+        let deferred: Q.Deferred<string> = Q.defer();
+        fs.readFile(ymlFilePath, { encoding: "utf8" }, async (error, data) => {
+
+            if (error) {
+                vscode.window.showErrorMessage(error.message);
+            }
+            else {
+                let fileContent = Mustache.render(data, context);
+                let ymlFileUri = vscode.Uri.file(vscode.workspace.workspaceFolders[0].uri.path + "/azure-pipelines.yml");
+                fs.writeFileSync(ymlFileUri.fsPath, fileContent);
+                await vscode.workspace.saveAll(true);
+                deferred.resolve(ymlFileUri.fsPath);
+            }
+        });
+
+        return deferred.promise;
+    }
+
+    /**
+     * commits yaml pipeline file into the local repo and pushes the commit to remote branch.
+     * @param pipelineYamlPath : local path of yaml pipeline in the repository
+     * @returns: thenable object which resolves once commit is pushed to remote branch, and failure message if unsuccessful
+     */
+    public async commitAndPushPipelineFile(pipelineYamlPath: string): Promise<{ commitId: string, branch: string }> {
+        // TODO
+        // unstages the changes already staged
+        // stages the file at the mentioned path
+        // commit the staged file
+        // stage the previously staged changes
+
+
+        await this.gitReference.add(pipelineYamlPath);
+        let commit = await this.gitReference.commit("Add yml file to workspace", pipelineYamlPath);
+        let status = await this.gitReference.status();
+        let remote = status.tracking;
+        let branch = status.current;
+        if (remote && branch) {
+            remote = remote.substr(0, remote.indexOf(branch) - 1);
+            await this.gitReference.push(remote, branch);
+        }
+        else {
+            throw new Error("Cannot add yml file to your git repository, remote is not set");
+        }
+        return {
+            branch: branch,
+            commitId: commit.commit
+        };
+    }
+
+    private async getGitRepoDetails(azureDevOps: AzureDevOpsService): Promise<GitRepoDetails> {
+        let status = await this.gitReference.status();
+        let branch = status.current;
+        let tracking = status.tracking.substr(0, status.tracking.indexOf(branch) - 1);
+        let remoteUrl = await this.gitReference.remote(["get-url", tracking]);
+
+        if (remoteUrl) {
+            if (remoteUrl.indexOf(GithubUrl) >= 0) {
+                //https://github.com/dikhakha/DemoNodeApp.git
+
+                let repoId = remoteUrl.substring(GithubUrl.length, remoteUrl.indexOf(".git"));
+                return <GitRepoDetails> {
+                    sourceProvider: SourceProviderType.Github,
+                    repositoryId: repoId,
+                    repositoryName: repoId,
+                    branch: "",
```

Comment:
we are not using branch and commitId until the end after all selection is made. should we change the return type here or probably remove these 2 fields from return type ? I think repo type,id and name is good enough here.

Reply:
Added branch and commitId as we may not create a commit to checkin file, if user already has a pipeline file in the repo. In that case we will just use the current branch and its latest commit.